crates/syn2mas/src/synapse_reader/checks.rs

```diff
@@ -0,0 +1,290 @@
+    }
+
+    (warnings, errors)
+}
+
+/// Check that the given Synapse configuration is sane for migration to a MAS with the given MAS configuration.
+///
+/// # Errors
+///
+/// - If any necessary section of MAS config cannot be parsed.
+/// - If the MAS password configuration (including any necessary secrets) can't be loaded.
+pub async fn synapse_config_check_against_mas_config(
+    synapse: &Config,
+    mas: &Figment,
+) -> Result<(Vec<CheckWarning>, Vec<CheckError>), Error> {
+    let mut errors = Vec::new();
+    let warnings = Vec::new();
+
+    let mas_passwords = PasswordsConfig::extract_or_default(mas)?;
+    let mas_password_schemes = mas_passwords
+        .load()
+        .await
+        .map_err(Error::MasPasswordConfig)?;
+
+    let mas_matrix = MatrixConfig::extract(mas)?;
+
+    // Look for the MAS password hashing scheme that will be used for imported Synapse passwords,
+    // then check the configuration matches so that Synapse passwords will be compatible with MAS.
+    let mut found_password_scheme = false;
+    for (scheme_version, scheme_algorithm, _scheme_rounds, scheme_secret) in &mas_password_schemes {
+        if *scheme_version == MIGRATED_PASSWORD_VERSION {
+            if scheme_algorithm != &PasswordAlgorithm::Bcrypt {
+                errors.push(CheckError::PasswordSchemeNotBcrypt);
+            }
+
+            let synapse_pepper = synapse
+                .password_config
+                .pepper
+                .as_ref()
+                .map(String::as_bytes);
+            if scheme_secret.as_deref() != synapse_pepper {
+                errors.push(CheckError::PasswordSchemeWrongPepper);
+            }
+
+            found_password_scheme = true;
+        }
+    }
+
+    if !found_password_scheme {
+        errors.push(CheckError::MissingPasswordScheme);
+    }
+
+    if synapse.allow_guest_access {
+        errors.push(CheckError::GuestsEnabled);
+    }
+
+    if synapse.server_name != mas_matrix.homeserver {
+        errors.push(CheckError::ServerNameMismatch {
+            synapse: synapse.server_name.clone(),
+            mas: mas_matrix.homeserver.clone(),
+        });
+    }
+
+    Ok((warnings, errors))
+}
+
+/// Check that the Synapse database is sane for migration. Returns a list of warnings and errors.
+///
+/// # Errors
+///
+/// - If there is some database connection error, or the given database is not a Synapse database.
+/// - If the OAuth2 section of the MAS configuration could not be parsed.
+#[tracing::instrument(skip_all)]
+pub async fn synapse_database_check(
+    synapse_connection: &mut PgConnection,
+    synapse: &Config,
+    mas: &Figment,
+) -> Result<(Vec<CheckWarning>, Vec<CheckError>), Error> {
+    #[derive(FromRow)]
+    struct UpstreamOAuthProvider {
+        auth_provider: String,
+        num_users: i64,
+    }
+
+    let mut errors = Vec::new();
+    let warnings = Vec::new();
+
+    let num_guests: i64 = query_scalar("SELECT COUNT(1) FROM users WHERE is_guest <> 0")
```

Comment:
I wonder whether we should disable inactive guests here? Like the server may have historical guests but they don't matter? That would be the case on m.org for example

Reply:
This so far is a copy of the existing tool for this aspect. My intention here is to disable this error once we import the bare minimum guest metadata that lets us not clash with guest IDs in the future?
Whichever way you look at it, we need to remove this error before we do matrix.org :-)

Reply:
Alright, let's tackled that in another PR :)